rust/src/mqtt/detect.rs

```diff
@@ -0,0 +1,481 @@
+    tx: &MQTTTransaction,
+    username: u8,
+    password: u8,
+    will: u8,
+    will_retain: u8,
+    clean_session: u8,
+) -> u8 {
+    for msg in tx.msg.iter() {
+        if let MQTTOperation::CONNECT(ref cv) = msg.op {
+            let mut ok = true;
+            match MQTTFlagState::from(username) {
+                MQTTFlagState::MUST_BE_SET => {
+                    if !cv.username_flag {
+                        ok = false;
+                    }
+                },
+                MQTTFlagState::CANT_BE_SET => {
+                    if cv.username_flag {
+                        ok = false;
+                    }
+                },
+                _ => {}
+            }
+            match MQTTFlagState::from(password) {
+                MQTTFlagState::MUST_BE_SET => {
+                    if !cv.password_flag {
+                        ok = false;
+                    }
+                },
+                MQTTFlagState::CANT_BE_SET => {
+                    if cv.password_flag {
+                        ok = false;
+                    }
+                },
+                _ => {}
+            }
+            match MQTTFlagState::from(will) {
+                MQTTFlagState::MUST_BE_SET => {
+                    if !cv.will_flag {
+                        ok = false;
+                    }
+                },
+                MQTTFlagState::CANT_BE_SET => {
+                    if cv.will_flag {
+                        ok = false;
+                    }
+                },
+                _ => {}
+            }
+            match MQTTFlagState::from(will_retain) {
+                MQTTFlagState::MUST_BE_SET => {
+                    if !cv.will_retain {
+                        ok = false;
+                    }
+                },
+                MQTTFlagState::CANT_BE_SET => {
+                    if cv.will_retain {
+                        ok = false;
+                    }
+                },
+                _ => {}
+            }
+            match MQTTFlagState::from(clean_session) {
+                MQTTFlagState::MUST_BE_SET => {
+                    if !cv.clean_session {
+                        ok = false;
+                    }
+                },
+                MQTTFlagState::CANT_BE_SET => {
+                    if cv.clean_session {
+                        ok = false;
+                    }
+                },
+                _ => {}
+            }
+            if ok {
+                return 1;
+            }
+        }
+    }
+    return 0;
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn rs_mqtt_tx_get_connect_clientid(
+    tx: &MQTTTransaction,
+    buffer: *mut *const u8,
+    buffer_len: *mut u32,
+) -> u8 {
+    for msg in tx.msg.iter() {
+        if let MQTTOperation::CONNECT(ref cv) = msg.op {
+            let p = &cv.client_id;
+            if p.len() > 0 {
+                *buffer = p.as_ptr();
+                *buffer_len = p.len() as u32;
+                return 1;
+            }
```

Comment:
Are we sure we will have only one `CONNECT` in one transaction ? (and thus not multiple buffers to match)

Reply:
The spec only specifies one CONNECT (http://docs.oasis-open.org/mqtt/mqtt/v3.1.1/csprd02/mqtt-v3.1.1-csprd02.html#_Toc385349764) in a compliant communication: 

> A Client can only send the CONNECT Packet once over a Network Connection. The    
> Server MUST process a second CONNECT Packet sent from a Client as a protocol
> violation and disconnect the Client

Reply:
Ok, thanks